graphbolt/src/shared_memory_utils.cc

```diff
@@ -0,0 +1,248 @@
+      int64_t size = tensor.numel() * tensor.element_size();
+      memcpy(mem_buf, tensor.data_ptr(), size);
+      mem_buf = static_cast<char*>(mem_buf) + size;
+    }
+  }
+  return shm;
+}
+
+/**
+ * @brief Load tensors data from shared memory.
+ * @param name The name of shared memory.
+ * @param tensor_metas The meta info of tensors, including tensor shape and
+ * dtype.
+ * @param optional_tensor_metas The meta info of optional tensors, including a
+ * flag indicating whether the optional tensor has value, tensor shape and
+ * dtype.
+ *
+ * @return A tuple of shared memory holing the tensors and optional
+ * tensors.
+ */
+static std::tuple<
+    SharedMemoryPtr, std::vector<torch::Tensor>,
+    std::vector<torch::optional<torch::Tensor>>>
+LoadTensorsDataFromSharedMemory(
+    const std::string& name,
+    const std::vector<std::pair<std::vector<int64_t>, torch::ScalarType>>&
+        tensor_metas,
+    const std::vector<
+        std::tuple<bool, std::vector<int64_t>, torch::ScalarType>>&
+        optional_tensor_metas) {
+  auto shm = std::make_unique<SharedMemory>(name);
+  int64_t memory_size = 0;
+  for (const auto& meta : tensor_metas) {
+    int64_t size = std::accumulate(
+        meta.first.begin(), meta.first.end(), 1, std::multiplies<int64_t>());
+    memory_size += size * torch::elementSize(meta.second);
+  }
+  for (const auto& meta : optional_tensor_metas) {
+    if (std::get<0>(meta)) {
+      int64_t size = std::accumulate(
+          std::get<1>(meta).begin(), std::get<1>(meta).end(), 1,
+          std::multiplies<int64_t>());
+      memory_size += size * torch::elementSize(std::get<2>(meta));
+    }
+  }
+  auto mem_buf = shm->Open(memory_size);
+  std::vector<torch::Tensor> tensors;
+  for (const auto& meta : tensor_metas) {
+    auto tensor = torch::from_blob(mem_buf, meta.first, meta.second);
+    tensors.push_back(tensor);
+    int64_t size = std::accumulate(
+        meta.first.begin(), meta.first.end(), 1, std::multiplies<int64_t>());
+    mem_buf =
+        static_cast<char*>(mem_buf) + size * torch::elementSize(meta.second);
+  }
+  std::vector<torch::optional<torch::Tensor>> optional_tensors;
+  for (const auto& meta : optional_tensor_metas) {
+    if (std::get<0>(meta)) {
+      auto tensor =
+          torch::from_blob(mem_buf, std::get<1>(meta), std::get<2>(meta));
+      optional_tensors.push_back(tensor);
+      int64_t size = std::accumulate(
+          std::get<1>(meta).begin(), std::get<1>(meta).end(), 1,
+          std::multiplies<int64_t>());
+      mem_buf = static_cast<char*>(mem_buf) +
+                size * torch::elementSize(std::get<2>(meta));
+    } else {
+      optional_tensors.push_back(torch::nullopt);
+    }
+  }
+  return std::make_tuple(
+      std::move(shm), std::move(tensors), std::move(optional_tensors));
+}
+
+SharedMemoryTensors CopyTensorsToSharedMemory(
+    const std::string& name, const std::vector<torch::Tensor>& tensors,
```

Comment:
make every tensor optional can save a lot of code.

Reply:
OK